util/misc.go

```diff
@@ -53,3 +57,46 @@ func GetStack() []byte {
 	buf = buf[:stackSize]
 	return buf
 }
+
+// WithRecovery wraps goroutine startup call with recovery handle.
+// There are two out of box common handlers `WithPanicLogf` and `WithPanicMetricInc`,
+// Caller can custom define new handler use last optional parameters.
+func WithRecovery(ctx context.Context, fn func(ctx context.Context), handlerOpts ...func(r interface{})) {
+	if len(handlerOpts) == 0 {
+		handlerOpts = append(handlerOpts, DefaultLogPanicHandler)
+	}
+	defer func() {
+		r := recover()
+		if r != nil {
+			fmt.Println()
+		}
+		for _, handler := range handlerOpts {
```

Comment:
it should be called when `r != nil`?

Reply:
at first, check `r != nil` too, but find it's hard to use to https://github.com/pingcap/tidb/pull/7666/files#diff-4576a60d1fcb8da2eed7999714e054baL526

it always `close(innerFinished)` but if `r != nil`  need send a error though `innerFinished` before close channel.

so, at last always call to handler, and give handler to check `r`

Reply:
Let handler check the panic makes the code looks ugly. And it also makes the API hard to use.

Reply:
handler don't check, `withRecovery` maybe need another optional "finally" handler.

let handler check is more flexible, and caller doesn't often self-define handler, and just combine predefine handlers

Reply:
seems we are handcrafting a `try ... catch ... finally ...` exception handler.

Reply:
I think the good point is our `catch-finally` can be reusable and easy check some `go  xxx()` not recovery.